Adapter/PhpcrOdmAdapter.php

```diff
@@ -190,12 +212,49 @@ public function getReferringAutoRoutes($contentDocument)
     public function findRouteForUri($uri, UriContext $uriContext)
     {
         $path = $this->getPathFromUri($uri);
+        $document = $this->dm->find(null, $path);
 
-        return $this->dm->find(null, $path);
+        if ($document instanceof AutoRouteInterface) {
+            return $document;
+        }
+
+        return null;
     }
 
     private function getPathFromUri($uri)
     {
         return $this->baseRoutePath . $uri;
     }
+
+    /**
+     * Convert the given generic document to an auto route document, migrating any children which it may have.
+     *
+     * @param Generic $document
+     * @param object $contentDocument
+     * @param string $autoRouteTag
+     * @param string $routeType
+     * @return AutoRouteInterface
+     */
+    private function migrateGenericToAutoRoute(Generic $document, $contentDocument, $autoRouteTag, $routeType)
+    {
+        $autoRouteClassName = $this->autoRouteFqcn;
+        $mapper = $this->dm->getConfiguration()->getDocumentClassMapper();
+        $mapper->writeMetadata($this->dm, $document->getNode(), $autoRouteClassName);
+        $this->dm->getPhpcrSession()->save();
+        // Detach is needed to force Doctrine to re-load the node
+        $this->dm->detach($document);
+        $autoRoute = $this->dm->find(null, $document->getId());
+
+        if (!$autoRoute instanceof $autoRouteClassName) {
+            throw new \RuntimeException(
+                'Something went wrong converting Generic node into an AutoRouteInterface node.'
```

Comment:
Done [here](https://github.com/mmenozzi/RoutingAutoBundle/blob/670ed28373116d684074f5aa6b7354008163f949/Tests/Unit/Adapter/PhpcrOdmAdapterTest.php#L133). What do you think? It's too much tied to the implementation?